### PR TITLE
Merge latest changes from dev into master

### PR DESCRIPTION
**Pull request by richardwu**
We need to update master with dev now that dev is stable. Changes in this merge:

- Fixed numerous exception handling bugs
- Normalization of clean and training data is now consistent 
- Added Travis CI
- Added a unit/integration test
- Fixed a couple of typos in README
- Added `create_db_ubuntu.sh` and `create_pd_user_ubuntu.sh` scripts for initializing postgres
- Added numerous docstrings and comments to functions
- Introduced a posterior probability estimator to approximate the conditional distribution of values for weak labelling and domain generation. This improves weak supervised learning. See parameters `weak_label_thresh`, `domain_prune_thresh`, and `max_domain` (#43)
- Outputting weights for featurizers when `print_fw=True`
- Disable pool creation when `threads <= 1`
- Fixed UTF-8 encoding issue when ingesting training data
- Fixed bug where non-lowercase attribute names caused Postgres to fail
- Added two new test datasets `met` and `Adult`

### CODE REVIEW

File: dataset/dataset.py [hunk 1]
```diff
@@ -37,6 +19,11 @@ class AuxTables(Enum):
     inf_values_idx = 6
     inf_values_dom = 7
 
+class CellStatus(Enum):
```
Comment (minafarid) [inline dataset/dataset.py:23]: Can we please add some documentation about this Enum and 1 line about what each value means

File: dataset/dataset.py [hunk 2]
```diff
@@ -153,32 +174,43 @@ def generate_aux_table_sql(self, aux_table, query, index_attrs=False):
...
         return vid
 
     def get_statistics(self):
+        """
+        get_statistics returns:
+            1. self.total_tuples (total # of tuples)
+            2. self.single_attr_stats ({ attribute -> { value -> count } })
+              the frequency (# of entities) of a given attribute-value
+            3. self.pair_attr_stats ({ attr1 -> { attr2 -> {val1 -> {val2 -> count } } } })
+              where DataFrame contains 3 columns:
```
Comment (minafarid) [inline dataset/dataset.py]: I think this is no longer a dataframe

File: dataset/dataset.py [hunk 3]
```diff
@@ -188,10 +220,10 @@ def get_statistics(self):
     def collect_stats(self):
         """
         collect_stats memoizes:
-          1. self.single_attr_stats ({ attribute -> Series (value -> count) })
+          1. self.single_attr_stats ({ attribute -> { value -> count } })
             the frequency (# of entities) of a given attribute-value
-          2. self.pair_attr_stats ({ attr1 -> { attr2 -> DataFrame } } where
-            DataFrame contains 3 columns:
+          2. self.pair_attr_stats ({ attr1 -> { attr2 -> {val1 -> {val2 -> count } } } })
+            where DataFrame contains 3 columns:
```
Comment (minafarid) [inline dataset/dataset.py:235]: Again, I think that the returned stats are not in a dataframe but rather the dictionary that is described in the previous line.

File: domain/domain.py [hunk 4]
```diff
@@ -97,44 +97,44 @@ def setup_attributes(self):
...
+            { attr1 -> { attr2 -> {val1 -> {val2 -> count } } } } where
                 DataFrame contains 3 columns:
-                  <attr1>: all possible values for attr1 ('val1')
-                  <attr2>: all values for attr2 that appeared at least once with <val1> ('val2')
-                  <count>: frequency (# of entities) where attr1: val1 AND attr2: val2
+                  <val1>: all possible values for attr1
+                  <val2>: all values for attr2 that appeared at least once with <val1>
+                  <count>: frequency (# of entities) where attr1: <val1> AND attr2: <val2>
 
         to a flattened 4-level dictionary { attr1 -> { attr2 -> { val1 -> [Top K list of val2] } } }
```
Comment (minafarid) [inline domain/domain.py]: No dataframes, and this no longer flattens, it just filters values above `tau`

File: domain/domain.py [hunk 5]
```diff
@@ -57,16 +59,14 @@ def find_correlations(self):
...
+        for attr in df.columns:
+            df[attr] = df[attr].astype('category').cat.codes
         # drop columns with only one value and tid column
-        d = d.loc[:, (d != 0).any(axis=0)]
-        d = d.drop(['_tid_'], axis=1)
+        df = df.loc[:, (df != 0).any(axis=0)]
         # Computer correlation across attributes
-        m_corr = d.corr()
+        m_corr = df.corr()
         self.correlations = m_corr
```
Comment (minafarid) [inline domain/domain.py]: merge the previous two lines?

File: domain/domain.py [hunk 6]
```diff
@@ -259,14 +339,13 @@ def get_domain_cell(self, attr, row):
...
-                    domain.update(set(candidates))
+                    domain.update(candidates)
                 except KeyError as missing_val:
-                    if pd.isnull(row[attr]):
-                        pass
-                    else:
-                        # error since co-occurrence must be at least 1 (since the
-                        # current row counts as one co-occurrence).
-                        raise Exception(missing_val)
+                    if not pd.isnull(row[attr]):
```
Comment (minafarid) [inline domain/domain.py]: can we revert the logic to `continue` if `pd.isnull` to make it more readable

File: README.md [hunk 7]
```diff
@@ -14,125 +19,147 @@ predictions, and insights form noisy, incomplete, and erroneous data.
...
 ```bash
-$ conda create -n holo_env python=2.7
+$ conda create -n hc36 python=3.6
 ```
 
 Upon starting/restarting your terminal session, you will need to activate your
 conda environment by running
 ```bash
-$ source activate holo_env
+$ source activate hc36
```
Comment (jonmio) [inline README.md]: Consider adding instructions to deactivate venv when finished with `source deactivate`

File: dataset/dataset.py [hunk 8]
```diff
@@ -65,44 +59,65 @@ def __init__(self, name, env):
...
         load_data takes a CSV file of the initial data, adds tuple IDs (_tid_)
         to each row to uniquely identify an 'entity', and generates unique
         index numbers for each attribute/column.
 
         Creates a table with the user supplied 'name' parameter (e.g. 'hospital').
-        """
 
+        :param name: (str) name to initialize dataset with.
+        :param fpath: (str) filepath to CSV file.
+        :param na_values: (str) value that identifies a NULL value
```
Comment (jonmio) [inline dataset/dataset.py:73]:
```suggestion
        :param na_values: (List[str] or str) value that identifies a NULL value.
```

File: dataset/dataset.py [hunk 9]
```diff
@@ -65,44 +59,65 @@ def __init__(self, name, env):
...
+
+            # Use '_nan_' to represent NULL values
+            df.fillna('_nan_', inplace=True)
+
+            logging.info("Loaded %d rows with %d cells", self.raw_data.df.shape[0], self.raw_data.df.shape[0] * self.raw_data.df.shape[1])
+
             # Call to store to database
             self.raw_data.store_to_db(self.engine.engine)
-            status = 'DONE Loading '+f_name
+            status = 'DONE Loading {fname}'.format(fname=os.path.basename(fpath))
```
Comment (jonmio) [inline dataset/dataset.py:110]: Nit: Seems like both str.format() and % are used to format strings in the codebase

File: dataset/dataset.py [hunk 10]
```diff
@@ -153,56 +172,72 @@ def generate_aux_table_sql(self, aux_table, query, index_attrs=False):
...
+        return self.raw_data.df
 
     def get_attributes(self):
-        if self.raw_data:
-            attrs = self.raw_data.get_attributes()
-            attrs.remove('_tid_')
-            return attrs
-        else:
+        """
+        get_attributes return the trainable/learnable attributes (i.e. exclude meta
```
Comment (jonmio) [inline dataset/dataset.py:183]:
```suggestion
        get_attributes returns the trainable/learnable attributes (i.e. exclude meta
```

File: dataset/dataset.py [hunk 11]
```diff
@@ -153,56 +172,72 @@ def generate_aux_table_sql(self, aux_table, query, index_attrs=False):
...
 
     def collect_stats(self):
         """
         collect_stats memoizes:
-          1. self.single_attr_stats ({ attribute -> Series (value -> count) })
+          1. self.single_attr_stats ({ attribute -> { value -> count } })
             the frequency (# of entities) of a given attribute-value
-          2. self.pair_attr_stats ({ attr1 -> { attr2 -> DataFrame } } where
-            DataFrame contains 3 columns:
+          2. self.pair_attr_stats ({ attr1 -> { attr2 -> {val1 -> {val2 -> count } } } })
```
Comment (jonmio) [inline dataset/dataset.py:234]: I think descriptions for a few keys in the dict are missing. 
Copied from `get_statistics` docstring
```suggestion
               3. self.pair_attr_stats ({ attr1 -> { attr2 -> {val1 -> {val2 -> count } } } })
              the statistics for each pair of attributes, attr1 and attr2, where:
                <attr1>: first attribute
                <attr2>: second attribute
                <val1>: all values of <attr1>
                <val2>: values of <attr2> that appear at least once with <val1>.
                <count>: frequency (# of entities) where attr1=val1 AND attr2=val2
        """
```

File: dataset/dbengine.py [hunk 12]
```diff
@@ -98,22 +72,81 @@ def execute_query(self, query):
...
         tic = time.clock()
         conn = self.engine.connect()
         result = conn.execute(stmt)
         conn.close()
         toc = time.clock()
         logging.debug('Time to create index: %.2f secs', toc-tic)
         return result
+
+    def _apply_func(self, func, collection):
+        if self._pool is None:
```
Comment (jonmio) [inline dataset/dbengine.py:107]: Why set self._pool to None when pool size is 1 instead of 1? Is it to avoid overhead incurred by Python multiprocessing library?

File: dataset/dbengine.py [hunk 13]
```diff
@@ -98,22 +72,81 @@ def execute_query(self, query):
...
+def _execute_query(args, conn_args):
+    query_id = args[0]
+    query = args[1]
+    logging.debug("Starting to execute query %s with id %s", query, query_id)
+    tic = time.clock()
+    con = psycopg2.connect(conn_args)
+    cur = con.cursor()
+    cur.execute(query)
+    res = cur.fetchall()
+    con.close()
```
Comment (jonmio) [inline dataset/dbengine.py:121]: Consider using with keyword in Python to handle db connection?

File: dataset/table.py [hunk 14]
```diff
@@ -1,78 +1,88 @@
...
-        else:
-            raise Exception("Empty Dataframe associated with table "+self.name+". Cannot return attributes.")
+        """
+        get_attributes returns the columns that are trainable/learnable attributes
+        (i.e. exclude meta-columns like _tid_).
+        """
+        if self.df.empty:
+            raise Exception("Empty Dataframe associated with table {name}. Cannot return attributes.".format(
+                name=self.name))
+        return list(col for col in self.df.columns if col not in self.exclude_attr_cols)
```
Comment (jonmio) [inline dataset/table.py:85]: Could just use list comprehension directly if a list is being returned.
```suggestion
        return [col for col in self.df.columns if col not in self.exclude_attr_cols]
```

File: dataset/dbengine.py [hunk 15]
```diff
@@ -98,22 +72,81 @@ def execute_query(self, query):
...
         logging.debug('Time to create index: %.2f secs', toc-tic)
         return result
+
+    def _apply_func(self, func, collection):
+        if self._pool is None:
+            return list(map(func, collection))
+        return self._pool.map(func, collection)
+
+
+def _execute_query(args, conn_args):
```
Comment (jonmio) [inline dataset/dbengine.py:112]: Are docstrings needed for private fns?

File: dcparser/dcparser.py [hunk 16]
```diff
@@ -4,47 +4,45 @@
...
-
 class Parser:
     """
     This class creates interface for parsing denial constraints
     """
-
     def __init__(self, env, dataset):
         """
         Constructing parser interface object
+
```
Comment (jonmio) [inline dcparser/dcparser.py:15]: Params in docstring seem out of date

File: domain/domain.py [hunk 17]
```diff
@@ -143,29 +141,34 @@ def get_active_attributes(self):
...
             raise Exception("No attribute contains erroneous cells.")
         return set(itertools.chain(*result))
 
-    def get_corr_attributes(self, attr):
+    def get_corr_attributes(self, attr, thres):
         """
         get_corr_attributes returns attributes from self.correlations
         that are correlated with attr with magnitude at least self.cor_strength
         (init parameter).
+
```
Comment (jonmio) [inline domain/domain.py:189]: Missing :param attr:

File: domain/domain.py [hunk 18]
```diff
@@ -201,25 +207,119 @@ def generate_domain(self):
...
+            # Assign weak label if domain value exceeds our weak label threshold
+            weak_label, weak_label_prob = max(preds, key=lambda pred: pred[1])
+
+            if weak_label_prob >= self.weak_label_thresh:
+                num_weak_labels+=1
+
+                weak_label_idx = domain_values.index(weak_label)
+                row['weak_label'] = weak_label
+                row['weak_label_idx'] = weak_label_idx
+                row['fixed'] = CellStatus.WEAK_LABEL.value
```
Comment (jonmio) [inline domain/domain.py:358]: At this point in the domain df, there may be rows where `weak_label` is set to some non-null value, but these cells might not be used as labelled data in the repair model because the `weak_label_prob` doesn't exceed `weak_label_threshold`. 

AFAICT this shouldn't cause any problems in the repair model because a cell is considered labelled training data if `fixed` is set to `CellStatus.WEAK_LABEL.value`. However, it might make more sense to set the weak_label for these values to null or some invalid value to avoid confusion.

File: domain/estimator.py [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+    """
+    Estimator is an abstract class for posterior estimators that estimate
+    the posterior of p(value | other values) for the purpose of domain generation
+    and weak labelling.
+    """
+    __metaclass__ = ABCMeta
+
+    def __init__(self, dataset):
+        """
+        :param dataset: (Dataset)
```
Comment (jonmio) [inline domain/estimator.py:15]: Params in docstring out of date.

File: domain/estimators/naive_bayes.py [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+        for val1 in values:
+            val1_count = self._freq[attr][val1]
+            log_prob = math.log(float(val1_count)/float(self._n_tuples))
+            correlated_attributes = self._get_corr_attributes(attr)
+            total_log_prob = 0.0
+            for at in correlated_attributes:
+                if at != attr:
+                    val2 = row[at]
+                    val2_count = self._freq[at][val2]
+                    val2_val1_count = 0.1
```
Comment (jonmio) [inline domain/estimators/naive_bayes.py]: Could you leave a comment on what this variable is and why 0.1?

File: domain/estimators/recurrent_logistic.py [hunk 21]
```diff
@@ -0,0 +1,378 @@
...
+    RecurrentLogistic is an Estimator that approximates posterior of
+    p(v_cur | v_init) by training a logistic regression model to predict the current
+    value in a cell given all other initial values using features
+    of the other initial values such as co-occurrence.
+    """
+
+    def __init__(self, dataset, pruned_domain, active_attrs):
+        """
+        :param dataset: (Dataset) original dataset
+        :param pruned_domain: (dict) of tid -> attr -> value for the domain
```
Comment (jonmio) [inline domain/estimators/recurrent_logistic.py]: I think pruned_domain is a Dict where the key is tid (string)  and value is a Dict where the key is attr(string) and the value is a list of domain values (List[str])

File: domain/estimators/recurrent_logistic.py [hunk 22]
```diff
@@ -0,0 +1,378 @@
...
+        """
+        Reinitialize featurizers that depend on updated current values.
+
+        :param reuse_stats: (bool) reuses frequency and co-occurrence statistics from
+            raw Dataset instead of re-computing.
+        """
+
+        # Featurizers used for training.
+        freq, cooccur_freq = None, None
+        if reuse_stats:
```
Comment (jonmio) [inline domain/estimators/recurrent_logistic.py]: Has the data changed in the raw dataset? Why might we want to recompute stats here?

File: domain/estimators/recurrent_logistic.py [hunk 23]
```diff
@@ -0,0 +1,378 @@
...
+        :param num_epochs: (int) number of epochs PER recurrent iteration.
+        :param batch_size: (int) size of batch in stochastic gradient descent.
+        """
+        batch_losses = []
+        for recur_idx in range(1, num_recur + 1):
+            # We need to update our statistics in featurizers on the 2nd and
+            # later iteration with our newest current values.
+            if recur_idx > 1:
+                self._update_featurizers()
+            self._update_training_data()
```
Comment (jonmio) [inline domain/estimators/recurrent_logistic.py]: Why is it necessary to update training data after each iteration? Does training mutate the training data?

File: repair/featurize/constraintfeat.py [hunk 24]
```diff
@@ -38,28 +58,23 @@ def gen_feat_tensor(violations, total_vars, classes):
 
 
 class ConstraintFeat(Featurizer):
```
Comment (jonmio) [inline repair/featurize/constraintfeat.py]: Would be nice to have a docstring for each featurizer describing either the intuition behind using these features, or what feature these featurizers are generating.

File: repair/featurize/freqfeat.py [hunk 25]
```diff
@@ -5,16 +5,13 @@
...
+        self.all_attrs = self.ds.get_attributes()
         self.attrs_number = len(self.ds.attr_to_idx)
         total, single_stats, pair_stats = self.ds.get_statistics()
         self.total = total
-        self.single_stats = {}
-        for attr in single_stats:
-            self.single_stats[attr] = single_stats[attr].to_dict()
+        self.single_stats = single_stats
 
     def gen_feat_tensor(self, input, classes):
```
Comment (jonmio) [inline repair/featurize/freqfeat.py:16]: gen_feat_tensor is sometimes an instance method in some featurizers and a normal function outside of the class in others.

File: tests/test_holoclean_repair.py [hunk 26]
```diff
@@ -0,0 +1,59 @@
...
+        infer_labeled = True
+
+        hc.repair_errors(featurizers, infer_labeled=infer_labeled)
+
+        # 5. Evaluate the correctness of the results.
+        hc.evaluate(fpath='../testdata/hospital_100_clean.csv',
+                    tid_col='tid',
+                    attr_col='attribute',
+                    val_col='correct_val',
+                    infer_labeled=infer_labeled)
```
Comment (jonmio) [inline tests/test_holoclean_repair.py]: Would it maybe make sense to assert that certain eval metrics like precision and recall fall within some interval so we can test that HoloClean's performance hasn't degraded significantly?

**Comment by jonmio** [comment]
Left some comments. Mostly style related and some documentation fixes. Also, left a few questions.

File: README.md (hunk 7 above)
Comment (minafarid) [reply README.md]: Done
